Avoid overlap of point and axis hover labels

dagroe opened the pull request:
This is an attempt at resolving issue #3973 

When hover info extends below the lower edge of the plot it may overlap the hover axis info, as illustrated in this codepen: https://codepen.io/dagroe/pen/BaPmNXr

The goal of this MR is to move the hover info up to always be above the lower edge and thus above the axis hover info.

dagroe commented:
Before I fix the tests, @archmoj could you have a look if this is a possible solution? 😄

archmoj commented:
Thanks for the PR.
Could you fix the error in the codepen please?

dagroe commented:
Sorry about that. I fixed it.

archmoj commented:
Thanks and here is a [codepen](https://codepen.io/MojtabaSamimi/pen/YzjYNVZ) showing the proposed behavior.
Looks good to my eyes.
@alexcjohnson what do you think?

archmoj commented on src/components/fx/hover.js:
```diff
@@ -1518,7 +1518,7 @@ function hoverAvoidOverlaps(hoverLabels, axKey, fullLayout) {
             size: d.by * (axIsX ? YFACTOR : 1) / 2,
             pmin: 0,
-            pmax: (axIsX ? fullLayout.width : fullLayout.height)
+            pmax: (axIsX ? fullLayout.width : fullLayout.height) - ax._mainLinePosition
```
Here one should also take into account the `side` of the axis (i.e. *top* & *left* / *right*?).
Right now for an `xaxis` with `side` set to *top*, the position is not correct: [demo](https://codepen.io/MojtabaSamimi/pen/gOjogee).

dagroe replied:
Yes, if the general approach of modifying the boundaries here is fine, I'll fix this and add tests for all cases (axis and sides).

alexcjohnson replied:
I like it! In addition to x/y and side let's only make this change if we're using one of the "compare" modes (`hovermode: 'x'|'y'`) - there can be cases with multiple labels even when `hovermode: 'closest'`, for example box plots.
<img width="664" alt="Screenshot 2023-01-23 at 17 12 12" src="https://user-images.githubusercontent.com/2678795/214161543-5d4bdfc8-a909-404b-ad0b-55e13e32f64b.png">

dagroe commented:
I updated my branch to cover the cases of `hovermode: 'y'` and different sides.

I looked into the failing tests. This fix results in less space for hover labels, thus in many cases fewer labels are displayed, which makes the test fail.
To me this seems to be a larger caveat, especially since in some cases there is no overlap  and the entire space could be used, e.g. here:
<img width="447" alt="Screenshot 2023-01-23 at 12 10 40" src="https://user-images.githubusercontent.com/6624011/214229002-f798abaa-4db2-4ffc-bc52-20bbc9a00b8d.png">

So my question is if that is a sacrifice we are willing to make or whether we should look into different approaches, e.g. actually checking for overlap with the axis label or moving the axis label to the side.

alexcjohnson commented:
That sacrifice is fine IMO - a common label as short as `1` like in that image is itself an edge case, in most cases there will be overlap. But it looks like all of those labels got pushed too far up, the lowest one could go right on the axis line.

dagroe commented:
I have adapted the approach to check for overlap of the hover label's background `rect` with the axis label. With that all existing tests pass.

<img width="498" alt="Screenshot 2023-01-31 at 12 15 13" src="https://user-images.githubusercontent.com/6624011/215745583-1489cc9f-472c-45c2-ad9f-20789e9bd80e.png">

 I have added one test to test the overlap functionality. 
<img width="498" alt="Screenshot 2023-01-31 at 12 15 19" src="https://user-images.githubusercontent.com/6624011/215745596-a02878a7-4360-4865-9c5d-5c6f1eb9fdb4.png">


> But it looks like all of those labels got pushed too far up, the lowest one could go right on the axis line.

It seems the algorithm layouts all labels and then only removes those that overlap the boundaries. It marks points as deleted while iterating but does not remove those points from the positioning process. I'd rather not touch this as part of this PR.

dagroe commented:
Unfortunately I cannot reproduce the failing test locally.

dagroe commented:
@archmoj Could you help me with the one failing test? Locally the test passes fine and I don't see an overlap of labels or why there would be one. Other than this test, this PR is ready from my side.

archmoj commented on src/components/fx/hover.js:
```diff
@@ -1662,6 +1734,39 @@ function hoverAvoidOverlaps(hoverLabels, axKey, fullLayout) {
+    return {
+        alignShift: alignShift,
+        xx: txx,
+        x2x: tx2x
```
Instead of using `xx` and `x2x`, could we possibly improve the naming here?

dagroe replied:
to be honest I am not 100% sure what the original author intended with variable names `txx` and `tx2x`. I guessed `t` might be translation or shift, so I kept the `xx` and `x2x` part with the `shift` part being in the function name. Do you have a suggestion for changing those? :)

dagroe replied:
I had another look at this. Seems `txx` was to mean `x` of text (`tx`) and `tx2x` `x` of text2 (`tx2`). I changed it to more descriptive variable names accordingly.

archmoj commented on src/components/fx/hover.js:
```diff
@@ -1675,38 +1780,27 @@ function alignHoverText(hoverLabels, rotateLabels, scaleX, scaleY) {
-        var offsetX = 0;
-        var offsetY = d.offset;
+        var shiftX = getLabelShiftX(d);
+        var offsets = getHoverLabelOffsets(d, rotateLabels);
```
Let's store `x` and `y` in variables and continue using `offsetX` and `offsetY` in the code below.
```js
var offset = getHoverLabelOffset(d, rotateLabels);
var offsetX = offset.x;
var offsetY = offset.y;
```

archmoj replied:
Would you mind considering my comment above: https://github.com/plotly/plotly.js/pull/6442/files#r1096117470
That would help reduce the diff and avoid multiple key access.
Thank you!

dagroe replied:
Sure. added changes.

archmoj commented:
> @archmoj Could you help me with the one failing test? Locally the test passes fine and I don't see an overlap of labels or why there would be one. Other than this test, this PR is ready from my side.

The timezone test is to ensure we produce correct hover in different timezones: 
https://github.com/plotly/plotly.js/blob/70c49ce593501e37d9dbf3b28de2caeec95a6b01/.circleci/config.yml#L52-L68
Could you switch the timezone or a use docker to debug?

dagroe commented:
> Could you switch the timezone or a use docker to debug?

I can try. The same test also fails in `no-gl-jasmine`, so I thought timezone is not the issue.

archmoj commented:
> > Could you switch the timezone or a use docker to debug?
> 
> I can try. The same test also fails in `no-gl-jasmine`, so I thought timezone is not the issue.

You are right. It's also `no-gl-jasmine`. It passes on my machine too.
So let's add `noCI` to the test i.e.
```js
it('@noCI centered-aligned, should stack nicely upon each other', function(done) {
```

dagroe commented:
Ok. I'll give docker a try on Monday. If it also passes there I'll add the `@noCI`.

archmoj commented on src/components/fx/hover.js:
```diff
@@ -829,9 +830,10 @@ function _hover(gd, evt, subplot, noHoverEvent, eventTarget) {
 
     if(!helpers.isUnifiedHover(hovermode)) {
-        hoverAvoidOverlaps(hoverLabels, rotateLabels ? 'xa' : 'ya', fullLayout);
+        hoverAvoidOverlaps(hoverLabels, rotateLabels, fullLayout, hoverText.commonLabel);
```
Looks like a good catch! 
Just to double check `'xa' : 'ya'` part was not needed here and `rotateLabels` is the argument that should be passed to the function.

dagroe commented:
I got a docker container up based on CI config. The test also failed there.

I had to dig a bit but ultimately the positions and sizes on CI had minor differences to the local run. In one case this led to an overlap of less than 1 pixel between one hover label and axis label while the other labels didn't overlap. Thus the algorithm produced different layouts on CI and local runs.

I changed the code to ignore overlaps of less than 1 pixel.

So the approach is not perfect and we can still have some overlap in edge cases, e.g. when not all hover labels are stacked on top of each other, but I assume that is an edge case that we can live with.

archmoj commented:
Thanks @dagroe. Now it looks very good to my eyes.
@alexcjohnson over to you :pray:

dagroe commented:
@alexcjohnson would appreciate your review to hopefully see this in a release soon 😄

alexcjohnson commented:
looks like in that particular case, the 2-digit labels slide behind the 1-digit labels 🤔

dagroe commented:
Good catch @alexcjohnson. Seems I was actually using the wrong `rect` behind `text2` instead of the actual `path` behind `text`. This caused different `x` position for 1-digit and 2-digit labels which resulted in two `pointgroups` instead of one. Apparently there is no overlap prevention between different `pointgroups` but I assume when they are built correctly then there should be no overlap.

I fixed it and added a test for the example you provided.

No idea what is happening with CI. @archmoj can you help with that?

alexcjohnson commented:
hmph that's a weird CI error, but I get a similar failure when I try to pull your changes locally, looks like something went wrong at the git level 🙄 
```
> git pull dagroe issue-3973-overlap-hover
remote: fatal: bad tree object 59144c5f100ab5bde6fbcd473e71f3a75fa060ec
remote: aborting due to possible repository corruption on the remote side.
fatal: protocol error: bad pack header
```
Usually we don't like to do this because it makes re-reviews harder, but maybe you could try to rebase locally, then force push? Just so git goes through the motions again and hopefully avoids whatever it broke here?

dagroe commented:
That worked, thanks.

One test I added fails. Seems the intersection takes the entire axis-label width into account instead only the actual `path` behind the label. I just looked at the calculation to create the `path`, but this will require a bit of time to make sure it's correct in all cases, so I'll do it on Monday.

alexcjohnson reviewed (approved):
Behavior looks good to me now - I tried a few more cases like several separated groups that get pushed together:

```js
Plotly.react(
    gd,
    [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20].map(v0=>{
        const v = v0 > 9 ? v0 + 100 : v0; return {x:[100,200,300],y:[v,v+1,v+2]}
    }),
    {width:500,height:400,hovermode:'x',margin:{t:0}}
)
```

Seems like the logic is solid now. Once the tests are passing I think this is good to go! 💃

dagroe commented:
I implemented the changes mentioned above. The test that failed passes now as expected but one test marked as `flaky` fails. It passes locally. Maybe you could re-run that one pipeline if that makes sense?

archmoj commented:
> I implemented the changes mentioned above. The test that failed passes now as expected but one test marked as `flaky` fails. It passes locally. Maybe you could re-run that one pipeline if that makes sense?

All passed.
If possible, please create `draftlog/6442_fix.md` file as described here: https://github.com/plotly/plotly.js/blob/master/draftlogs/README.md
Thank you!

dagroe commented:
> All passed. If possible, please create `draftlog/6442_fix.md` file as described here: https://github.com/plotly/plotly.js/blob/master/draftlogs/README.md Thank you!

👍 Done

aberres commented:
@alexcjohnson  @archmoj Is this ready to be merged?
